network_simulator/host.py

```diff
@@ -23,4 +27,65 @@ def __init__(self, address=None, flows={}, flow_packets_received={},
         self.flow_packets_received = flow_packets_received
         self.link = link
 
-    # TODO(team): More event subclasses to come?
\ No newline at end of file
+
+class HostReceivedPacketEvent(Event):
+    """
+    Host receives packets from a flow.
+    """
+
+    def __init__(self, host, packet):
+        """
+        Sanity check: make sure destination was this Host and that this isn't
+        a RouterPacket.
+        :ivar host: host that receives packet
+        :ivar packet: packet received
+        """
+        Event.__init__(self)
+
+        # Sanity check
+        assert packet.dest_id == host and \
+        packet.packet_id != ROUTER_PACKET_DEFAULT_ID
+
+        self.host = host
+        self.packet = packet
+
+    def run(self, main_event_loop, statistics):
+        """
+        Put packet in flow_packets_received and sort.
+        Update statistics.
+        :param main_event_loop: event loop where new Events will be scheduled.
+        :param statistics: statistics to update
+        """
+        # add packet to host.flow_packets_received
+        self.flow_packets_received[self.packet.flow_id] = self.packet.packet_id
+        self.flow_packets_received.sort()
+        
+        statistics.host_packet_received(self.host, self.packet,
+                                        main_event_loop.global_clock_sec)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        If packet was Ack, look up Flow and say packet was received.
+        Schedule FlowReceivedAckEvent.
+        Populate AckPacket for selective repeat.
+
+        If packet was data packet, generate Ack and schedule via
+        DeviceToLink event unless packet already received earlier.
+
+        Ackpackets, set original data packet's start time for RTT.
+        AckPacket ID same as DataPacket ID
+        :param main_event_loop: schedule new Events
+        :param statistics: update statistics
+        """
+
+        if isinstance(self.packet, AckPacket):
+            main_event_loop.schedule_new_event_with_delay(self, self.flow,
+                                                          self.packet)
+            ack_flow_id = self.packet.flow_id
+            for i in self.flow_packets_received:
+                if self.flow_packets_received[i].flow_id == ack_flow_id:
+                    self.flow_packets_received[i].packet_id = self.packet_id
+        if isinstance(self.packet, DataPacket) and \
+                        self.packet_id not in self.flow_packets_received:
+            a_packet = AckPacket(self.packet)
```

Comment:
For the AckPacket construction, look at the constructor you wrote in packet.py. Many of the values default to None. We do not want that to happen in this case.

You need to send in the following parameters into that constructor.
* packet_id = self.packet.packet_id, so the AckPacket has the same ID as the DataPacket that triggered it. This is important since we need to know "what" is actually being acknowledged as having been received.
* flow_id = self.packet.flow_id. ACKs need to have the same Flow ID as their triggering DataPacket so that we can track them properly when they are received.
* source_id = self.host. This AckPacket is originating from this Host.
* dest_id = self.packet.source_id. You are sending this Packet back to where it came from.
* start_time_sec = main_event_loop.global_clock_sec. This is the current time at which this ACK was sent back.
* flow_packets_received = self.flow_packets_received[self.flow_id]. This is the sorted list of Packets received for the given Flow, and this is also what you updated in run(). We need to send this because Flows will use this information to figure out if there are any gaps in the list of Packets. I have implemented this logic in FlowReceivedAckEvent's find_missing_packets() function, if you want to refer to it.
* data_packet_start_time_sec = self.packet.start_time_sec. This is the time at which the DataPacket was generated. We need to store this in the ACK so that we can calculate the round trip time when this ACK makes it back to the Host it came from.

After you have constructed the AckPacket, you need to make a DeviceToLinkEvent object. This constructor will take the ACK packet, the link (which is self.host.link, not self.link as you have written here), and the destination device "dest_dev". The destination device here is the *other* end of the Link (Links have two ends). In other words, it is the end of the Link that is not the Host. You can find this with the convenience function get_other_end() in link.py (take a look at this). The way you would do this is "link = self.host.link" and then "link.get_other_end(self.host)", to find whatever's on the other end of this Link.

Once you've constructed the DeviceToLinkEvent object, schedule it immediately with main_event_loop.schedule_event_with_delay(...) as specified earlier. What this Event will do is just add the Packet to the Link's buffer, so it's fine to schedule it immediately.